src/ui/control/geolocate_control.js

```diff
@@ -72,24 +97,139 @@ class GeolocateControl extends Evented {
     }
 
     onRemove() {
+        // clear the geolocation watch if exists
+        if (this._geolocationWatchID !== undefined) {
+            window.navigator.geolocation.clearWatch(this._geolocationWatchID);
+            this._geolocationWatchID = undefined;
+        }
+
+        // clear the marker from the map
+        if (this.options.showUserLocation) {
+            this._userLocationDotMarker.remove();
+        }
+
         this._container.parentNode.removeChild(this._container);
         this._map = undefined;
     }
 
     _onSuccess(position) {
-        this._map.jumpTo({
-            center: [position.coords.longitude, position.coords.latitude],
-            zoom: 17,
-            bearing: 0,
-            pitch: 0
-        });
+        if (this.options.trackUserLocation) {
+            // keep a record of the position so that if the state is BACKGROUND and the user
+            // clicks the button, we can move to ACTIVE_LOCK immediately without waiting for
+            // watchPosition to trigger _onSuccess
+            this._lastKnownPosition = position;
+
+            switch (this._watchState) {
+            case 'WAITING_ACTIVE':
+            case 'ACTIVE_LOCK':
+            case 'ACTIVE_ERROR':
+                this._watchState = 'ACTIVE_LOCK';
+                this._geolocateButton.classList.remove('mapboxgl-ctrl-geolocate-waiting');
+                this._geolocateButton.classList.remove('mapboxgl-ctrl-geolocate-active-error');
+                this._geolocateButton.classList.add('mapboxgl-ctrl-geolocate-active');
+                break;
+            case 'BACKGROUND':
+            case 'BACKGROUND_ERROR':
+                this._watchState = 'BACKGROUND';
+                this._geolocateButton.classList.remove('mapboxgl-ctrl-geolocate-waiting');
+                this._geolocateButton.classList.remove('mapboxgl-ctrl-geolocate-background-error');
+                this._geolocateButton.classList.add('mapboxgl-ctrl-geolocate-background');
+                break;
+            default:
+                assert(false, `Unexpected watchState ${this._watchState}`);
+            }
+            this._geolocateButton.classList.remove('mapboxgl-ctrl-geolocate-waiting');
```

Comment:
Nit: I think this line makes the equivalent lines in the `case`s above redundant.